### PR TITLE
Add VIA keymap for Zinc

### PR DESCRIPTION
**Pull request by monksoffunk**
## Description

Add VIA support for Zinc

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by monksoffunk** [comment]
Is there anything I should do for this PR?

File: keyboards/zinc/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                if(keymap_config.swap_lalt_lgui==false){
+                    register_code(KC_LANG1);
+                } else {
+                    SEND_STRING(SS_LALT("`"));
+                }
+            } else {
+                unregister_code(KC_LANG1);
+            }
+            return false;
+            break;
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/keymap.c:163]:
```suggestion
```

File: keyboards/zinc/keymaps/via/rules.mk [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+COMMAND_ENABLE = no        # Commands for debug and configuration
+NKRO_ENABLE = no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no      # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.
+SWAP_HANDS_ENABLE = no        # Enable one-hand typing
+VIA_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/rules.mk:18]:
```suggestion
VIA_ENABLE = yes
EXTRAKEY_ENABLE = yes
```

File: keyboards/zinc/keymaps/via/rules.mk [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# Zinc keyboard customize
+LED_BACK_ENABLE = no        # LED backlight (Enable SK6812mini backlight)
+LED_UNDERGLOW_ENABLE = no   # LED underglow (Enable WS2812 RGB underlight)
+LED_BOTH_ENABLE = yes        # LED backlight and underglow
+LED_RGB_CONT = yes           # LED continuous backlight or/and underglow between left Zinc and right Zinc
+RGB_MATRIX = no             # RGB LED Matrix
+RGB_MATRIX_SPLIT_RIGHT = no # RGB Matrix for RIGHT Hand
+LED_ANIMATIONS = yes        # LED animations
+IOS_DEVICE_ENABLE = no      # connect to IOS device (iPad,iPhone)
+Link_Time_Optimization = no # if firmware size over limit, try this option
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/rules.mk:40]:
```suggestion
LTO_ENABLE = no # if firmware size over limit, try this option
```

File: keyboards/zinc/keymaps/via/rules.mk [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+endif
+
+ifeq ($(strip $(Link_Time_Optimization)),yes)
+  EXTRAFLAGS += -flto -DUSE_Link_Time_Optimization
+endif
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
+
+
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/rules.mk:133]:
```suggestion
```

File: keyboards/zinc/keymaps/via/rules.mk [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+endif
+
+ifeq ($(strip $(RGB_MATRIX_SPLIT_RIGHT)), yes)
+    OPT_DEFS += -DRGB_MATRIX_SPLIT_RIGHT
+endif
+
+ifeq ($(strip $(Link_Time_Optimization)),yes)
+  EXTRAFLAGS += -flto -DUSE_Link_Time_Optimization
+endif
+
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/rules.mk:129]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/zinc/keymaps/via/keymap.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+enum custom_keycodes {
+    QWERTY = SAFE_RANGE,
+    KANA,
+    EISU,
+    ADJUST,
+    RGBRST
+};
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = { \
```
Comment (fauxpark) [inline keyboards/zinc/keymaps/via/keymap.c:56]:
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by monksoffunk** [comment]
@fauxpark thanks for your review!!

File: keyboards/zinc/keymaps/via/keymap.c [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    return true;
+}
+
+
+void matrix_init_user(void) {
+#ifdef RGBLIGHT_ENABLE
+    rgblight_init();
+    RGB_current_config = rgblight_config;
+#endif
+}
```
Comment (drashna) [inline keyboards/zinc/keymaps/via/keymap.c:181]: Or you could do this: 
```suggestion
void keyboard_post_init_user(void) {
#ifdef RGBLIGHT_ENABLE
    RGB_current_config = rgblight_config;
#endif
}
```

File: keyboards/zinc/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+layer_state_t layer_state_set_user(layer_state_t state) {
+#ifdef RGBLIGHT_LAYERS
+    // Both layers will light up if both kb layers are active
+    rgblight_set_layer_state(1, layer_state_cmp(state, 1));
+    rgblight_set_layer_state(2, layer_state_cmp(state, 2));
+#endif
+    return update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST2);
+}
```
Comment (drashna) [inline keyboards/zinc/keymaps/via/keymap.c:190]: You want to update the layer state before you check the state. 

```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
    state = update_tri_layer_state(state, _LOWER, _RAISE, _ADJUST2);
#ifdef RGBLIGHT_LAYERS
    // Both layers will light up if both kb layers are active
    rgblight_set_layer_state(1, layer_state_cmp(state, 1));
    rgblight_set_layer_state(2, layer_state_cmp(state, 2));
#endif
    return state;
}
```

File: keyboards/zinc/keymaps/via/keymap.c [hunk 9]
```diff
@@ -0,0 +1,195 @@
+#include QMK_KEYBOARD_H
+
+#ifdef RGBLIGHT_ENABLE
+//Following line allows macro to read current RGB settings
+extern rgblight_config_t rgblight_config;
+rgblight_config_t RGB_current_config;
```
Comment (drashna) [inline keyboards/zinc/keymaps/via/keymap.c:6]: These don't actually seem to be used in any real way. 

```suggestion
```

File: keyboards/zinc/keymaps/via/keymap.c [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+            } else {
+                unregister_code(KC_LANG1);
+            }
+            return false;
+        case RGBRST:
+#ifdef RGBLIGHT_ENABLE
+            if (record->event.pressed) {
+                eeconfig_update_rgblight_default();
+                rgblight_enable();
+                RGB_current_config = rgblight_config;
```
Comment (drashna) [inline keyboards/zinc/keymaps/via/keymap.c:167]:
```suggestion
```

File: keyboards/zinc/keymaps/via/keymap.c [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    }
+    return true;
+}
+
+
+void keyboard_post_init_user(void) {
+#ifdef RGBLIGHT_ENABLE
+    RGB_current_config = rgblight_config;
+#endif
+}
```
Comment (drashna) [inline keyboards/zinc/keymaps/via/keymap.c:180]:
```suggestion
```

File: keyboards/zinc/keymaps/via/keymap.c (hunk 9 above)
Comment (monksoffunk) [reply keyboards/zinc/keymaps/via/keymap.c]: Yes, these are old codes.
Thanks!

**Comment by monksoffunk** [comment]
Let me know if there is anything else I need to update.

**Comment by monksoffunk** [comment]
@fauxpark Please let me know if there is anything else I need to update.

**Review by noroadsleft — CHANGES_REQUESTED** [review]

File: keyboards/zinc/keymaps/via/keymap.c [hunk 12]
```diff
@@ -0,0 +1,187 @@
+#include QMK_KEYBOARD_H
```
[the remark this answers is not in the document]
Comment (monksoffunk) [reply keyboards/zinc/keymaps/via/keymap.c:17]: done!

**Review by drashna — APPROVED** [review]

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Comment by monksoffunk** [comment]
Thanks everyone!

**Pull request merged**